File: src/train.py
import os
import asyncio
import numpy as np
import tensorflow as tf
from dotenv import load_dotenv
from src.models.question_model import QuestionModel
from src.services.openai_service import OpenAIService
from src.data.questions import questions

async def generate_training_data(num_samples=1000):
    # Generujeme náhodné tréningové dáta
    X = np.random.rand(num_samples, 150)  # 150 otázok
    y = np.random.randint(0, 20, num_samples)  # 20 možných otázok
    return X, y

async def main():
    # Načítame premenné prostredia
    load_dotenv()
    
    # Inicializujeme služby
    model = QuestionModel()
    openai_service = OpenAIService()
    
    # Skúsime načítať existujúci model
    await model.load_model()
    
    # Testujeme OpenAI API
    if not openai_service.test_api_key():
        print('Chyba: OpenAI API kľúč nie je platný')
        return
    
    print('Začínam trénovanie modelu...')
    
    # Generujeme tréningové dáta
    X_train, y_train = await generate_training_data()
    
    # Trénujeme model
    history = model.model.fit(
        X_train,
        tf.keras.utils.to_categorical(y_train, num_classes=20),
        epochs=50,
        batch_size=32,
        validation_split=0.2
    )
    
    # Uložíme model
    await model.save_model()
    
    print('Trénovanie dokončené')
    print(f'Finálna presnosť: {history.history["accuracy"][-1]:.2f}')

if __name__ == '__main__':
    asyncio.run(main()) 

File: src/train.js
require('dotenv').config();
const QuestionModel = require('./models/QuestionModel');
const OpenAIService = require('./services/openaiService');
const { questions } = require('./data/questions');
const axios = require('axios');

class ModelTrainer {
    constructor() {
        this.model = new QuestionModel();
        this.openaiService = new OpenAIService();
        this.maxQuestionsPerSession = 20;
    }

    async initialize() {
        console.log('Inicializácia trénovania...');
        
        // Overenie OpenAI API
        try {
            await this.openaiService.testAPIKey();
        } catch (error) {
            console.error('Chyba pri overovaní OpenAI API:', error.message);
            process.exit(1);
        }

        // Overenie modelu
        const modelLoaded = await this.model.loadModel();
        if (!modelLoaded) {
            console.log('Vytváram nový model...');
        }

        // Overenie otázok
        if (questions.length === 0) {
            console.error('Žiadne otázky nie sú definované v questions.js');
            process.exit(1);
        }

        console.log(`Načítaných ${questions.length} otázok`);
    }

    async train() {
        try {
            await this.initialize();
            console.log('Začínam trénovanie modelu...');

            while (this.model.consecutiveSuccesses < this.model.requiredSuccesses) {
                console.log(`\nZačínam novú reláciu (${this.model.consecutiveSuccesses + 1}/${this.model.requiredSuccesses})`);
                
                const sessionSuccess = await this.runTrainingSession();
                
                if (sessionSuccess) {
                    this.model.consecutiveSuccesses++;
                    console.log(`Relácia úspešná! Celkový počet úspešných relácií: ${this.model.consecutiveSuccesses}`);
                    
                    // Uložíme model po každej úspešnej relácii
                    const saved = await this.model.saveModel();
                    if (!saved) {
                        console.error('Chyba pri ukladaní modelu po úspešnej relácii');
                        this.model.consecutiveSuccesses = 0;
                        continue;
                    }
                } else {
                    this.model.consecutiveSuccesses = 0;
                    console.log('Relácia neúspešná, začínam od začiatku.');
                }
            }

            console.log('\nTrénovanie dokončené! Ukladám finálny model...');
            const saved = await this.model.saveModel();
            if (!saved) {
                throw new Error('Chyba pri ukladaní finálneho modelu');
            }
            console.log('Model bol úspešne uložený.');
        } catch (error) {
            console.error('Kritická chyba pri trénovaní:', error);
            process.exit(1);
        }
    }

    async runTrainingSession() {
        const previousAnswers = [];
        let sessionSuccess = true;

        for (let i = 0; i < this.maxQuestionsPerSession; i++) {
            const availableQuestions = this.getAvailableQuestions(previousAnswers);
            if (availableQuestions.length === 0) {
                console.log('Žiadne dostupné otázky pre ďalšiu iteráciu');
                return false;
            }

            const selectedQuestion = await this.model.selectNextQuestion(previousAnswers, availableQuestions);
            
            try {
                const isValid = await this.openaiService.validateQuestionSelection(selectedQuestion, previousAnswers);
                
                if (!isValid) {
                    sessionSuccess = false;
                    break;
                }

                // Simulujeme odpoveď
                const answer = this.generateSimulatedAnswer(selectedQuestion);
                previousAnswers.push(answer);
            } catch (error) {
                console.error('Chyba pri validácii otázky:', error);
                sessionSuccess = false;
                break;
            }
        }

        return sessionSuccess;
    }

    getAvailableQuestions(previousAnswers) {
        const answeredIds = previousAnswers.map(a => a.questionId);
        return questions.filter(q => !answeredIds.includes(q.id));
    }

    generateSimulatedAnswer(question) {
        let value;
        switch (question.type) {
            case 'yes_no':
                value = Math.random() > 0.5;
                break;
            case 'slider':
                value = Math.floor(Math.random() * (question.options.max - question.options.min + 1)) + question.options.min;
                break;
            case 'multiple_choice':
                value = Math.floor(Math.random() * question.options.length);
                break;
        }

        return {
            questionId: question.id,
            type: question.type,
            value: value
        };
    }
}

// Spustenie trénovania
const trainer = new ModelTrainer();
trainer.train().catch(error => {
    console.error('Kritická chyba:', error);
    process.exit(1);
}); 